File: src/python/nlp/validation_engine.py
"""
Validation engine for checking parameters and permissions
"""

from typing import Dict, List, Any, Optional
from .types import (
    IntentType, 
    QueryParameters, 
    ConversationContext, 
    ValidationRule, 
    ValidationResult, 
    ValidationError, 
    ValidationWarning, 
    ValidationSeverity
)

class ValidationEngine:
    def __init__(self):
        self.validation_rules: Dict[IntentType, List[ValidationRule]] = {}
        self._initialize_validation_rules()

    def _initialize_validation_rules(self):
        """Initialize validation rules for each intent type"""
        
        # Rules for list_records
        self.validation_rules[IntentType.LIST_RECORDS] = [
            ValidationRule(
                parameter='table',
                validation_type='required',
                error_message='El nombre de la tabla es requerido',
                validation_func=lambda value: value and isinstance(value, str) and value.strip()
            )
        ]

        # Rules for create_record
        self.validation_rules[IntentType.CREATE_RECORD] = [
            ValidationRule(
                parameter='table',
                validation_type='required',
                error_message='El nombre de la tabla es requerido',
                validation_func=lambda value: value and isinstance(value, str) and value.strip()
            ),
            ValidationRule(
                parameter='fields',
                validation_type='required',
                error_message='Se requiere al menos un campo para crear el registro',
                validation_func=lambda value: value and isinstance(value, dict) and len(value) > 0
            )
        ]

        # Rules for update_record
        self.validation_rules[IntentType.UPDATE_RECORD] = [
            ValidationRule(
                parameter='table',
                validation_type='required',
                error_message='El nombre de la tabla es requerido',
                validation_func=lambda value: value and isinstance(value, str) and value.strip()
            ),
            ValidationRule(
                parameter='record_id',
                validation_type='required',
                error_message='El ID del registro es requerido',
                validation_func=lambda value: value and isinstance(value, str) and value.strip()
            ),
            ValidationRule(
                parameter='fields',
                validation_type='required',
                error_message='Se requiere al menos un campo para actualizar',
                validation_func=lambda value: value and isinstance(value, dict) and len(value) > 0
            )
        ]

        # Rules for delete_record
        self.validation_rules[IntentType.DELETE_RECORD] = [
            ValidationRule(
                parameter='table',
                validation_type='required',
                error_message='El nombre de la tabla es requerido',
                validation_func=lambda value: value and isinstance(value, str) and value.strip()
            ),
            ValidationRule(
                parameter='record_id',
                validation_type='required',
                error_message='El ID del registro es requerido',
                validation_func=lambda value: value and isinstance(value, str) and value.strip()
            )
        ]

        # Rules for create_webhook
        self.validation_rules[IntentType.CREATE_WEBHOOK] = [
            ValidationRule(
                parameter='table',
                validation_type='required',
                error_message='El nombre de la tabla es requerido',
                validation_func=lambda value: value and isinstance(value, str) and value.strip()
            ),
            ValidationRule(
                parameter='webhook_config',
                validation_type='required',
                error_message='La configuración del webhook es requerida',
                validation_func=lambda value: value and isinstance(value, dict) and value.get('notificationUrl')
            )
        ]

        # Rules for upload_attachment
        self.validation_rules[IntentType.UPLOAD_ATTACHMENT] = [
            ValidationRule(
                parameter='table',
                validation_type='required',
                error_message='El nombre de la tabla es requerido',
                validation_func=lambda value: value and isinstance(value, str) and value.strip()
            ),
            ValidationRule(
                parameter='record_id',
                validation_type='required',
                error_message='El ID del registro es requerido',
                validation_func=lambda value: value and isinstance(value, str) and value.strip()
            ),
            ValidationRule(
                parameter='attachment_data',
                validation_type='required',
                error_message='Los datos del adjunto son requeridos',
                validation_func=lambda value: value and isinstance(value, dict) and value.get('url')
            )
        ]

        # Rules for batch_create_records
        self.validation_rules[IntentType.BATCH_CREATE_RECORDS] = [
            ValidationRule(
                parameter='table',
                validation_type='required',
                error_message='El nombre de la tabla es requerido',
                validation_func=lambda value: value and isinstance(value, str) and value.strip()
            ),
            ValidationRule(
                parameter='max_records',
                validation_type='conditional',
                error_message='El número de registros debe estar entre 1 y 10',
                validation_func=lambda value: (
                    True if value is None else  # Optional
                    (isinstance(value, int) and 1 <= value <= 10)
                )
            )
        ]

    async def validate(
        self, 
        intent: IntentType, 
        parameters: QueryParameters, 
        context: ConversationContext
    ) -> ValidationResult:
        """
        Validate parameters against rules for the given intent
        """
        rules = self.validation_rules.get(intent, [])
        errors: List[ValidationError] = []
        warnings: List[ValidationWarning] = []
        suggestions: List[str] = []

        # Validate specific rules
        for rule in rules:
            value = self._get_parameter_value(parameters, rule.parameter)
            
            if rule.validation_type == 'required' and not value:
                errors.append(ValidationError(
                    field=rule.parameter,
                    message=rule.error_message,
                    code=f'MISSING_{rule.parameter.upper()}',
                    severity=ValidationSeverity.ERROR
                ))
            elif rule.validation_type == 'conditional' and value is not None:
                if rule.validation_func and not rule.validation_func(value):
                    errors.append(ValidationError(
                        field=rule.parameter,
                        message=rule.error_message,
                        code=f'INVALID_{rule.parameter.upper()}',
                        severity=ValidationSeverity.ERROR
                    ))

        # Additional validations
        additional_validation = await self._perform_additional_validations(intent, parameters, context)
        errors.extend(additional_validation.errors)
        warnings.extend(additional_validation.warnings)
        suggestions.extend(additional_validation.suggestions)

        return ValidationResult(
            is_valid=len(errors) == 0,
            errors=errors,
            warnings=warnings,
            suggestions=suggestions
        )

    def _get_parameter_value(self, parameters: QueryParameters, param_name: str) -> Any:
        """Get parameter value from QueryParameters"""
        # Check direct parameters
        if hasattr(parameters, param_name):
            return getattr(parameters, param_name)
        
        # Check nested parameters
        if parameters.fields and param_name in parameters.fields:
            return parameters.fields[param_name]
        
        if parameters.webhook_config and param_name in parameters.webhook_config:
            return parameters.webhook_config[param_name]
        
        if parameters.attachment_data and param_name in parameters.attachment_data:
            return parameters.attachment_data[param_name]
        
        return None

    async def _perform_additional_validations(
        self, 
        intent: IntentType, 
        parameters: QueryParameters, 
        context: ConversationContext
    ) -> ValidationResult:
        """
        Perform additional validations specific to intent
        """
        errors: List[ValidationError] = []
        warnings: List[ValidationWarning] = []
        suggestions: List[str] = []

        # Validate webhook URLs
        if intent == IntentType.CREATE_WEBHOOK and parameters.webhook_config and parameters.webhook_config.get('notificationUrl'):
            url = parameters.webhook_config['notificationUrl']
            if not self._is_valid_url(url):
                errors.append(ValidationError(
                    field='webhook_config.notificationUrl',
                    message='La URL del webhook no es válida',
                    code='INVALID_WEBHOOK_URL',
                    severity=ValidationSeverity.ERROR
                ))

        # Validate attachment URLs
        if intent == IntentType.UPLOAD_ATTACHMENT and parameters.attachment_data and parameters.attachment_data.get('url'):
            url = parameters.attachment_data['url']
            if not self._is_valid_url(url):
                errors.append(ValidationError(
                    field='attachment_data.url',
                    message='La URL del adjunto no es válida',
                    code='INVALID_ATTACHMENT_URL',
                    severity=ValidationSeverity.ERROR
                ))

        # Validate batch sizes
        if 'batch' in intent.value and parameters.max_records:
            if parameters.max_records > 10:
                warnings.append(ValidationWarning(
                    field='max_records',
                    message='El tamaño del lote es mayor a 10, se recomienda usar un valor menor',
                    code='LARGE_BATCH_SIZE',
                    suggestion='Usar un tamaño de lote de 10 o menos para mejor rendimiento'
                ))

        # Validate table names
        if parameters.table:
            if len(parameters.table) > 50:
                errors.append(ValidationError(
                    field='table',
                    message='El nombre de la tabla es demasiado largo',
                    code='TABLE_NAME_TOO_LONG',
                    severity=ValidationSeverity.ERROR
                ))

            if not self._is_valid_table_name(parameters.table):
                errors.append(ValidationError(
                    field='table',
                    message='El nombre de la tabla contiene caracteres no válidos',
                    code='INVALID_TABLE_NAME',
                    severity=ValidationSeverity.ERROR
                ))

        # Validate record IDs
        if parameters.record_id:
            if not self._is_valid_record_id(parameters.record_id):
                errors.append(ValidationError(
                    field='record_id',
                    message='El ID del registro no es válido',
                    code='INVALID_RECORD_ID',
                    severity=ValidationSeverity.ERROR
                ))

        return ValidationResult(
            is_valid=len(errors) == 0,
            errors=errors,
            warnings=warnings,
            suggestions=suggestions
        )

    def _is_valid_url(self, url: str) -> bool:
        """Check if URL is valid"""
        try:
            from urllib.parse import urlparse
            parsed = urlparse(url)
            return parsed.scheme in ['http', 'https']
        except Exception:
            return False

    def _is_valid_table_name(self, name: str) -> bool:
        """Check if table name is valid"""
        import re
        return bool(re.match(r'^[a-zA-ZáéíóúñÑ][a-zA-Z0-9áéíóúñÑ\s_-]*$', name))

    def _is_valid_record_id(self, record_id: str) -> bool:
        """Check if record ID is valid"""
        import re
        return bool(re.match(r'^[a-zA-Z0-9]+$', record_id))

    async def validate_permissions(
        self, 
        intent: IntentType, 
        context: ConversationContext
    ) -> Dict[str, Any]:
        """
        Validate permissions based on context
        """
        # For now, assume we have permissions
        # In real implementation, check user permissions
        
        sensitive_intents = [
            IntentType.DELETE_RECORD,
            IntentType.DELETE_WEBHOOK,
            IntentType.DELETE_TABLE,
            IntentType.DELETE_FIELD,
            IntentType.CREATE_TABLE,
            IntentType.CREATE_FIELD
        ]

        if intent in sensitive_intents:
            # Here we would check specific user permissions
            return {'has_permission': True}

        return {'has_permission': True}

    def generate_suggestions(self, error: ValidationError, context: ConversationContext) -> List[str]:
        """
        Generate correction suggestions for common errors
        """
        suggestions = []

        if error.code == 'MISSING_TABLE':
            if context.mentioned_entities.tables:
                table_list = ', '.join(context.mentioned_entities.tables)
                suggestions.append(f'¿Quisiste decir una de estas tablas: {table_list}?')
            else:
                suggestions.append('Especifica el nombre de la tabla: "en la tabla [nombre]"')

        elif error.code == 'MISSING_RECORD_ID':
            if context.current_record_id:
                suggestions.append(f'¿Te refieres al registro {context.current_record_id}?')
            else:
                suggestions.append('Especifica el ID del registro: "registro [ID]"')

        elif error.code == 'INVALID_TABLE_NAME':
            suggestions.append('Usa solo letras, números, espacios, guiones y guiones bajos')

        elif error.code == 'INVALID_WEBHOOK_URL':
            suggestions.append('La URL debe comenzar con http:// o https://')

        elif error.code == 'LARGE_BATCH_SIZE':
            suggestions.append('Para mejor rendimiento, usa lotes de 10 o menos registros')

        return suggestions

File: src/typescript/app/nlp/validation-engine.ts
/**
 * Motor de validación para verificar parámetros y permisos
 */

import { IntentType, QueryParameters, ConversationContext, ValidationRule } from '../types/nlp';

export interface ValidationResult {
  isValid: boolean;
  errors: ValidationError[];
  warnings: ValidationWarning[];
  suggestions: string[];
}

export interface ValidationError {
  field: string;
  message: string;
  code: string;
  severity: 'error' | 'critical';
}

export interface ValidationWarning {
  field: string;
  message: string;
  code: string;
  suggestion?: string;
}

export class ValidationEngine {
  private validationRules: Map<IntentType, ValidationRule[]> = new Map();

  constructor() {
    this.initializeValidationRules();
  }

  private initializeValidationRules(): void {
    // Reglas para list_records
    this.validationRules.set('list_records', [
      {
        parameter: 'table',
        type: 'required',
        validation: (value) => value && typeof value === 'string' && value.trim().length > 0,
        errorMessage: 'El nombre de la tabla es requerido'
      }
    ]);

    // Reglas para create_record
    this.validationRules.set('create_record', [
      {
        parameter: 'table',
        type: 'required',
        validation: (value) => value && typeof value === 'string' && value.trim().length > 0,
        errorMessage: 'El nombre de la tabla es requerido'
      },
      {
        parameter: 'fields',
        type: 'required',
        validation: (value) => value && typeof value === 'object' && Object.keys(value).length > 0,
        errorMessage: 'Se requiere al menos un campo para crear el registro'
      }
    ]);

    // Reglas para update_record
    this.validationRules.set('update_record', [
      {
        parameter: 'table',
        type: 'required',
        validation: (value) => value && typeof value === 'string' && value.trim().length > 0,
        errorMessage: 'El nombre de la tabla es requerido'
      },
      {
        parameter: 'recordId',
        type: 'required',
        validation: (value) => value && typeof value === 'string' && value.trim().length > 0,
        errorMessage: 'El ID del registro es requerido'
      },
      {
        parameter: 'fields',
        type: 'required',
        validation: (value) => value && typeof value === 'object' && Object.keys(value).length > 0,
        errorMessage: 'Se requiere al menos un campo para actualizar'
      }
    ]);

    // Reglas para delete_record
    this.validationRules.set('delete_record', [
      {
        parameter: 'table',
        type: 'required',
        validation: (value) => value && typeof value === 'string' && value.trim().length > 0,
        errorMessage: 'El nombre de la tabla es requerido'
      },
      {
        parameter: 'recordId',
        type: 'required',
        validation: (value) => value && typeof value === 'string' && value.trim().length > 0,
        errorMessage: 'El ID del registro es requerido'
      }
    ]);

    // Reglas para create_webhook
    this.validationRules.set('create_webhook', [
      {
        parameter: 'table',
        type: 'required',
        validation: (value) => value && typeof value === 'string' && value.trim().length > 0,
        errorMessage: 'El nombre de la tabla es requerido'
      },
      {
        parameter: 'webhookConfig',
        type: 'required',
        validation: (value) => value && typeof value === 'object' && value.notificationUrl,
        errorMessage: 'La configuración del webhook es requerida'
      }
    ]);

    // Reglas para upload_attachment
    this.validationRules.set('upload_attachment', [
      {
        parameter: 'table',
        type: 'required',
        validation: (value) => value && typeof value === 'string' && value.trim().length > 0,
        errorMessage: 'El nombre de la tabla es requerido'
      },
      {
        parameter: 'recordId',
        type: 'required',
        validation: (value) => value && typeof value === 'string' && value.trim().length > 0,
        errorMessage: 'El ID del registro es requerido'
      },
      {
        parameter: 'attachmentData',
        type: 'required',
        validation: (value) => value && typeof value === 'object' && value.url,
        errorMessage: 'Los datos del adjunto son requeridos'
      }
    ]);

    // Reglas para batch_create_records
    this.validationRules.set('batch_create_records', [
      {
        parameter: 'table',
        type: 'required',
        validation: (value) => value && typeof value === 'string' && value.trim().length > 0,
        errorMessage: 'El nombre de la tabla es requerido'
      },
      {
        parameter: 'maxRecords',
        type: 'conditional',
        validation: (value) => {
          if (!value) return true; // Opcional
          return typeof value === 'number' && value > 0 && value <= 10;
        },
        errorMessage: 'El número de registros debe estar entre 1 y 10'
      }
    ]);
  }

  async validate(
    intent: IntentType, 
    parameters: QueryParameters, 
    context: ConversationContext
  ): Promise<ValidationResult> {
    const rules = this.validationRules.get(intent) || [];
    const errors: ValidationError[] = [];
    const warnings: ValidationWarning[] = [];
    const suggestions: string[] = [];

    // Validar reglas específicas
    for (const rule of rules) {
      const value = this.getParameterValue(parameters, rule.parameter);
      
      if (rule.type === 'required' && !value) {
        errors.push({
          field: rule.parameter,
          message: rule.errorMessage,
          code: `MISSING_${rule.parameter.toUpperCase()}`,
          severity: 'error'
        });
      } else if (rule.type === 'conditional' && value) {
        const isValid = await rule.validation(value);
        if (!isValid) {
          errors.push({
            field: rule.parameter,
            message: rule.errorMessage,
            code: `INVALID_${rule.parameter.toUpperCase()}`,
            severity: 'error'
          });
        }
      }
    }

    // Validaciones adicionales específicas por intención
    const additionalValidation = await this.performAdditionalValidations(intent, parameters, context);
    errors.push(...additionalValidation.errors);
    warnings.push(...additionalValidation.warnings);
    suggestions.push(...additionalValidation.suggestions);

    return {
      isValid: errors.length === 0,
      errors,
      warnings,
      suggestions
    };
  }

  private getParameterValue(parameters: QueryParameters, paramName: string): any {
    return parameters[paramName] || 
           (parameters.fields && parameters.fields[paramName]) ||
           (parameters.webhookConfig && parameters.webhookConfig[paramName]) ||
           (parameters.attachmentData && parameters.attachmentData[paramName]);
  }

  private async performAdditionalValidations(
    intent: IntentType,
    parameters: QueryParameters,
    context: ConversationContext
  ): Promise<ValidationResult> {
    const errors: ValidationError[] = [];
    const warnings: ValidationWarning[] = [];
    const suggestions: string[] = [];

    // Validar URLs de webhook
    if (intent === 'create_webhook' && parameters.webhookConfig?.notificationUrl) {
      const url = parameters.webhookConfig.notificationUrl;
      if (!this.isValidUrl(url)) {
        errors.push({
          field: 'webhookConfig.notificationUrl',
          message: 'La URL del webhook no es válida',
          code: 'INVALID_WEBHOOK_URL',
          severity: 'error'
        });
      }
    }

    // Validar URLs de adjuntos
    if (intent === 'upload_attachment' && parameters.attachmentData?.url) {
      const url = parameters.attachmentData.url;
      if (!this.isValidUrl(url)) {
        errors.push({
          field: 'attachmentData.url',
          message: 'La URL del adjunto no es válida',
          code: 'INVALID_ATTACHMENT_URL',
          severity: 'error'
        });
      }
    }

    // Validar tamaño de lotes
    if (intent.includes('batch_') && parameters.maxRecords) {
      if (parameters.maxRecords > 10) {
        warnings.push({
          field: 'maxRecords',
          message: 'El tamaño del lote es mayor a 10, se recomienda usar un valor menor',
          code: 'LARGE_BATCH_SIZE',
          suggestion: 'Usar un tamaño de lote de 10 o menos para mejor rendimiento'
        });
      }
    }

    // Validar nombres de tabla
    if (parameters.table) {
      if (parameters.table.length > 50) {
        errors.push({
          field: 'table',
          message: 'El nombre de la tabla es demasiado largo',
          code: 'TABLE_NAME_TOO_LONG',
          severity: 'error'
        });
      }

      if (!/^[a-zA-ZáéíóúñÑ][a-zA-Z0-9áéíóúñÑ\s_-]*$/.test(parameters.table)) {
        errors.push({
          field: 'table',
          message: 'El nombre de la tabla contiene caracteres no válidos',
          code: 'INVALID_TABLE_NAME',
          severity: 'error'
        });
      }
    }

    // Validar IDs de registro
    if (parameters.recordId) {
      if (!/^[a-zA-Z0-9]+$/.test(parameters.recordId)) {
        errors.push({
          field: 'recordId',
          message: 'El ID del registro no es válido',
          code: 'INVALID_RECORD_ID',
          severity: 'error'
        });
      }
    }

    return {
      isValid: errors.length === 0,
      errors,
      warnings,
      suggestions
    };
  }



  private isValidUrl(url: string): boolean {
    try {
      const urlPattern = /^https?:\/\/[^\s/$.?#].[^\s]*$/i;
      return urlPattern.test(url);
    } catch {
      return false;
    }
  }

  /**
   * Valida permisos basados en el contexto
   */
  async validatePermissions(
    intent: IntentType, 
    context: ConversationContext
  ): Promise<{ hasPermission: boolean; reason?: string }> {
    // Por ahora, asumimos que tenemos permisos
    // En una implementación real, verificaríamos los permisos del usuario
    
    const sensitiveIntents: IntentType[] = [
      'delete_record', 
      'delete_webhook', 
      'create_table', 
      'delete_table',
      'create_field',
      'delete_field'
    ];

    if (sensitiveIntents.includes(intent)) {
      // Aquí se podría verificar si el usuario tiene permisos específicos
      return { hasPermission: true };
    }

    return { hasPermission: true };
  }

  /**
   * Genera sugerencias de corrección para errores comunes
   */
  generateSuggestions(error: ValidationError, context: ConversationContext): string[] {
    const suggestions: string[] = [];

    switch (error.code) {
      case 'MISSING_TABLE':
        if (context.mentionedEntities.tables.length > 0) {
          suggestions.push(`¿Quisiste decir una de estas tablas: ${context.mentionedEntities.tables.join(', ')}?`);
        } else {
          suggestions.push('Especifica el nombre de la tabla: "en la tabla [nombre]"');
        }
        break;

      case 'MISSING_RECORD_ID':
        if (context.currentRecordId) {
          suggestions.push(`¿Te refieres al registro ${context.currentRecordId}?`);
        } else {
          suggestions.push('Especifica el ID del registro: "registro [ID]"');
        }
        break;

      case 'INVALID_TABLE_NAME':
        suggestions.push('Usa solo letras, números, espacios, guiones y guiones bajos');
        break;

      case 'INVALID_WEBHOOK_URL':
        suggestions.push('La URL debe comenzar con http:// o https://');
        break;

      case 'LARGE_BATCH_SIZE':
        suggestions.push('Para mejor rendimiento, usa lotes de 10 o menos registros');
        break;
    }

    return suggestions;
  }
}
